src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs

```diff
@@ -1275,7 +1275,21 @@ protected override long ExecuteCmd(DocumentsOperationContext context)
                     var lastTransactionMarker = 0;
                     var handledAttachmentStreams = new HashSet<Slice>(SliceComparer.Instance);
                     context.LastDatabaseChangeVector = context.LastDatabaseChangeVector ?? DocumentsStorage.GetDatabaseChangeVector(context);
-                    foreach (var item in _replicationInfo.ReplicatedItems)
+
+                    var tx = context.Transaction.InnerTransaction.LowLevelTransaction;
+                    var replicationItems = _replicationInfo.ReplicatedItems;
+                    tx.OnDispose += _ =>
```

Comment:
What if this fail for the second time as well? would we leak here?

Reply:
In this case, the `TransactionOperationsMerger` will set the task of the `MergedDocumentReplicationCommand` as failed and the task that waiting for it (`ReceiveSingleDocumentsBatch`) will throw and finally will release the whole context.
This is the same as it handles it now for untreated items.

Reply:
okay, make sense. Nice catch :)